substrate/frame/transaction-payment/asset-conversion-tx-payment/src/lib.rs

```diff
@@ -202,108 +250,153 @@ impl<T: Config> core::fmt::Debug for ChargeAssetTxPayment<T> {
 	}
 }
 
-impl<T: Config> SignedExtension for ChargeAssetTxPayment<T>
+impl<T: Config> TransactionExtensionBase for ChargeAssetTxPayment<T>
 where
 	T::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
-	BalanceOf<T>: Send + Sync,
+	BalanceOf<T>: Send + Sync + From<u64>,
 	T::AssetId: Send + Sync,
 {
 	const IDENTIFIER: &'static str = "ChargeAssetTxPayment";
-	type AccountId = T::AccountId;
-	type Call = T::RuntimeCall;
-	type AdditionalSigned = ();
+	type Implicit = ();
+
+	fn weight() -> Weight {
+		<T as Config>::WeightInfo::charge_asset_tx_payment_asset()
+	}
+}
+
+impl<T: Config, Context> TransactionExtension<T::RuntimeCall, Context> for ChargeAssetTxPayment<T>
+where
+	T::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
+	BalanceOf<T>: Send + Sync + From<u64>,
+	T::AssetId: Send + Sync,
+	<T::RuntimeCall as Dispatchable>::RuntimeOrigin: AsSystemOriginSigner<T::AccountId> + Clone,
+{
+	type Val = (
+		// tip
+		BalanceOf<T>,
+		// who paid the fee
+		T::AccountId,
+		// transaction fee
+		BalanceOf<T>,
+	);
 	type Pre = (
 		// tip
 		BalanceOf<T>,
 		// who paid the fee
-		Self::AccountId,
+		T::AccountId,
 		// imbalance resulting from withdrawing the fee
 		InitialPayment<T>,
 	);
 
-	fn additional_signed(&self) -> core::result::Result<(), TransactionValidityError> {
-		Ok(())
-	}
-
 	fn validate(
 		&self,
-		who: &Self::AccountId,
-		call: &Self::Call,
-		info: &DispatchInfoOf<Self::Call>,
+		origin: <T::RuntimeCall as Dispatchable>::RuntimeOrigin,
+		call: &T::RuntimeCall,
+		info: &DispatchInfoOf<T::RuntimeCall>,
 		len: usize,
-	) -> TransactionValidity {
-		use pallet_transaction_payment::ChargeTransactionPayment;
-		let (fee, _) = self.withdraw_fee(who, call, info, len)?;
+		_context: &mut Context,
+		_self_implicit: Self::Implicit,
+		_inherited_implication: &impl Encode,
+	) -> ValidateResult<Self::Val, T::RuntimeCall> {
+		let who = origin.as_system_origin_signer().ok_or(InvalidTransaction::BadSigner)?;
+		// Non-mutating call of `compute_fee` to calculate the fee used in the transaction priority.
+		let fee = pallet_transaction_payment::Pallet::<T>::compute_fee(len as u32, info, self.tip);
+		self.can_withdraw_fee(&who, call, info, fee)?;
```

Comment:
Because `validate` is supposed to be non-mutating in terms of state. Calling `withdraw_fee` instead of just ensuring we can withdraw would change the state and would potentially impact the flow of extensions further down the pipeline when validating and impact the preparation phase of previous extensions.

Reply:
I don't really see the advantage, but I'm fine with it.